fml/concurrent_message_loop.cc

```diff
@@ -159,15 +148,12 @@ void ConcurrentTaskRunner::PostTask(const fml::closure& task) {
     return;
   }
 
-  if (auto loop = weak_loop_.lock()) {
-    loop->PostTask(task);
-    return;
+  {
+    std::scoped_lock lock(weak_loop_mutex_);
+    if (weak_loop_) {
+      weak_loop_->PostTask(task);
```

Comment:
There may be undefined behavior if a method like `ConcurrentMessageLoop::PostTask` is called on one thread after the object has begun deletion on a different thread.

At the point where this call happens, another thread may have already dropped the last reference to the `ConcurrentTaskLoop`, invoked `delete`, and called into the `ConcurrentTaskLoop` destructor.

The `ConcurrentTaskLoop` destructor will then try to acquire the `weak_loop_mutex_`, so the destructor will not make any further progress before the `ConcurrentMessageLoop::PostTask` call completes.  However, I don't think it's guaranteed to be safe to execute the `ConcurrentMessageLoop::PostTask` call while the `ConcurrentMessageLoop` is in that state.

Reply:
> There may be undefined behavior if a method like ConcurrentMessageLoop::PostTask is called on one thread after the object has begun deletion on a different thread.

Yep, good observation.  In order for it to get undefined behavior `ConcurrentTaskLoop` would need virtual methods and a subclass.  That would mean that the subclass could already have been deallocated while another thread is executing PostTask.  `ConcurrentTaskLoop` doesn't have virtual methods but that's easy to add and introduce undefined behavior on accident so I've made the class `final` to avoid this.

The order of operations of deallocation is well defined.  At the point where we are clearing out the reference the `ConcurrentMessageLoop ` instance variables will still be valid and the memory for `this` will still be around.  Since it is final we can know this is the very first thing that will be called.

Let me know if you you think we are still missing something.